### ISSUE TITLE
Problem with psutil -- fresh install on Ubuntu 15 fails

### ISSUE THREAD

**Issue by Asparagirl** [ArchiveTeam/ArchiveBot#184]
Doing a fresh install on Ubuntu 15 and I get up to this step..

```
pip3 install --user -r pipeline/requirements.txt
```

...and I get:

```
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -g -fstack-protector-strong -Wformat -W$

psutil/_psutil_linux.c:12:20: fatal error: Python.h: No such file or directory

compilation terminated.

error: command 'x86_64-linux-gnu-gcc' failed with exit status 1

----------------------------------------
Cleaning up...
Command /usr/bin/python3.5 -c "import setuptools, tokenize;__file__='/tmp/pip-build-kgtc3zry/psutil/setup.py';exec(comp$
Exception information:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pip/basecommand.py", line 122, in main
    status = self.run(options, args)
  File "/usr/lib/python3/dist-packages/pip/commands/install.py", line 309, in run
    requirement_set.install(install_options, global_options, root=options.root_path)
  File "/usr/lib/python3/dist-packages/pip/req.py", line 1436, in install
    requirement.install(install_options, global_options, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/pip/req.py", line 707, in install
    cwd=self.source_dir, filter_stdout=self._filter_install, show_stdout=False)
  File "/usr/lib/python3/dist-packages/pip/util.py", line 716, in call_subprocess
    % (command_desc, proc.returncode, cwd))
pip.exceptions.InstallationError: Command /usr/bin/python3.5 -c "import setuptools, tokenize;__file__='/tmp/pip-build-k$
```

Doing `apt-get install python-dev` before running all this didn't help.  Looks like something may be up with `psutil`?

**Comment by yipdw** [comment]
Interesting.  I was able to install psutil 3.2.2 in a Python 3.4 virtualenv built in Ubuntu 15.10: https://gist.github.com/yipdw/23681b06241085df0638

`pip install -r requirements.txt` in another virtualenv on the same system worked fine: https://gist.github.com/yipdw/9e028bceb3d3b96202d2

There's also a `python3-dev` package in Ubuntu 15.10; I have that installed.  Maybe that has different contents than `python-dev`?

(Note also that Python 3.5 may cause problems with `wpull`; you may want to use Python 3.4.)

**Comment by Asparagirl** [comment]
I restarted my work on a fresh install of Ubuntu 14 and it installed fine.

Something's up with Ubuntu 15 (on Digital Ocean, anyway) -- I did try both ```python-dev``` and ```python3-dev``` on that system and it just wasn't happening.  So, sticking with Ubuntu 14.

**Comment by ivan** [comment]
Ubuntu 15.10 comes with python3.5 by default, so you might need `python3.5-dev` on 15.10 instead.  Not that wpull would work on python3.5, due to https://github.com/chfoo/wpull/issues/290 and https://github.com/haypo/trollius/issues/4

**Comment by 12As** [comment]
In my tests, it appears that the default Python installed in the `python3` package is 3.4.3.  However, `python3-pip` depends on and uses Python 3.5.  Therefore, when you try to install anything with pip, it goes looking for the 3.5 headers.  The solution seem to be to use `python3-venv` (documented here: https://docs.python.org/3/library/venv.html) or a third-party virtual environment.  Doing so before line 43 of the install instructions installs a copy of pip3 in the environment that uses Python 3.4.3 and not 3.5.

**Comment by yipdw** [comment]
It sounds like there's a couple of solutions:

1. install `python3.5`
2. use `python3-venv` (which will vendor a copy of `pip`)

Does this sound right?

**Comment by 12As** [comment]
Because of the issue mentioned by Ivan, I think the only real solution is to use the virtual environment.  Given the state of dependency management in Python, that may be a good idea anyway to prevent clobbering of packages installed by `apt-get` with packages installed by `pip3` or vice versa.

**Comment by ivan** [comment]
https://bugs.launchpad.net/ubuntu/+source/python-pip/+bug/1518023 is now fixed, so the `python3-pip` on Ubuntu 15.10 uses python3.4.

**Comment by falconkirtaran** [comment]
All these versions are old now, but the installation is working well on Ubuntu 17.04 with psutil 5.2.2 and python 3.5.3 (with no venv or anything like that).
